src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java

```diff
@@ -85,18 +83,24 @@ Object base() {
     // factories
 
     public static MemorySegment makeNativeSegment(long bytesSize, long alignmentBytes) {
-        long alignedSize = bytesSize;
-
-        if (alignmentBytes > MAX_ALIGN) {
-            alignedSize = bytesSize + (alignmentBytes - 1);
+        if (VM.isDirectMemoryPageAligned()) {
+            alignmentBytes = Math.max(alignmentBytes, nioAccess.pageSize());
         }
+        long alignedSize = alignmentBytes > MAX_ALIGN ?
```

Comment:
Minor comment, it might be a little clearer if `MAX_ALIGN` was renamed to `MALLOC_MAX_ALIGN`.